test/mongoose_subdomain_core_SUITE.erl

```diff
@@ -0,0 +1,567 @@
+    %% also try to add some domains second time, as this is also possible
+    %% during subdomain_registration
+    [RegisteredDomain1, RegisteredDomain2 | _] = AllDomains,
+    mongoose_subdomain_core:add_domain(?DYNAMIC_HOST_TYPE1, RegisteredDomain1),
+    mongoose_subdomain_core:add_domain(?DYNAMIC_HOST_TYPE1, RegisteredDomain2),
+    mongoose_subdomain_core:sync(),
+    ?assertEqualLists(AllExpectedSubDomains, get_all_subdomains()),
+    ?assertEqual(NumOfDomainsToRemove,
+                 meck:num_calls(mongoose_hooks, disable_subdomain, 2)),
+
+    no_collisions(),
+    meck:unload(mongoose_domain_core).
+
+prevents_subdomain_overriding(_Config) ->
+    %% there are three possible subdomain names collisions:
+    %%   1) different domain/subdomain_pattern pairs produce one and the same subdomain
+    %%   2) attempt to register the same FQDN subdomain for 2 different host types
+    %%   3) domain/subdomain_pattern pair produces the same subdomain name as another
+    %%      FQDN subdomain
+    %%
+    %% collisions of the first type can eliminated by allowing only one level subdomains,
+    %% e.g. ensuring that subdomain template corresponds to this regex "^[^.]*\.@HOST@$".
+    %%
+    %% collisions of the second type are less critical as they can be detected during
+    %% init phase - they result in {error, subdomain_already_exists} return code, so
+    %% modules can detect it and crash at ?MODULE:start/2
+    %%
+    %% third type is hard to resolve in automatic way. one of the options is to ensure
+    %% that FQDN subdomains don't start with the same "prefix" as subdomain patterns.
+    %%
+    %% it's good idea to create a metric for such collisions, so devops can set some
+    %% alarm and react on it.
+    Pattern1 = mongoose_subdomain_utils:make_subdomain_pattern("sub.@HOST@"),
+    Pattern2 = mongoose_subdomain_utils:make_subdomain_pattern("sub.domain.@HOST@"),
+    Pattern3 = mongoose_subdomain_utils:make_subdomain_pattern("sub.domain.fqdn"),
+    Handler = mongoose_packet_handler:new(?MODULE),
+    %%----------------------------------------------------------------
+    %% testing type #1 subdomain names collision + double registration
+    %%----------------------------------------------------------------
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                                Pattern1, Handler)),
+    ?assertEqual(ok, mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                                Pattern2, Handler)),
+    ?assertEqual({error, already_registered},
+                 mongoose_subdomain_core:register_subdomain(?DYNAMIC_HOST_TYPE1,
+                                                            Pattern2, Handler)),
+    mongoose_domain_core:insert(<<"test">>, ?DYNAMIC_HOST_TYPE1, dummy_src),
+    mongoose_domain_core:insert(<<"domain.test">>, ?DYNAMIC_HOST_TYPE1, dummy_src),
```

Comment:
Please add a comment that this does not change the subdomain because of the collision.